### PR TITLE
Add beam particles

### LINKED ISSUE #3050: Particle Effects/Beam Weapons FR: Ability to spawn particles in a tube along the length of a beam
I would like to be able to spawn particle effects in a tube, of adjustable radius, along the length of an active beam. This would allow for various new visual effects -- especially vapor that lingers after the beam dies.

Because this specifies a region in which the effect would spawn, it would need some sort of "density" parameter -- the probability that a given point within the tube will spawn the particle effect each frame.

For cone-type effects, the 'incoming' vector should be the direction in which the beam is being fired. (Direction control is necessary for effective use of laser-type particles, which are very relevant for this feature.)

### PR DESCRIPTION
**Pull request by Baezon**
Adds a new `SourceOriginType` for particle sources corresponding to the entire length of a beam. All the special beam handling definitely... *chafes* a little with the framework, but I'm confident that if beams are to be done in the framework this is the best way to do it. Key here is that when asked for the origin's position, it gives random positions along its length, allowing all the existing shape and volume effects to otherwise work the same. 

The random distribution along its length is also biased towards the start of the beam, linearly decaying to 0 at its maximum range. This is done so that the distribution of particles essentially matches the opacity of the beam graphic itself, which also linearly decays to 0 towards its end.

Closes #3050

### CODE REVIEW

File: code/particle/ParticleSource.h [hunk 1]
```diff
@@ -65,6 +66,9 @@ class SourceOrigin {
 
 	/**
 	 * @brief Gets the current, global position of the origin
+	 * 
+	 * Be aware for beam soruces this will give *random* points along its length
```
Comment (Goober5000) [inline code/particle/ParticleSource.h:70]: "sources"

**Comment by EatThePath** [comment]
Would it be much extra work to add particles to warmup and warmdown states? I would have uses for it.

**Comment by Baezon** [comment]
Handling warmup and warmdown correctly is something I've already given some thought to, and definitely something I'd want too, but I decided to leave it off this commit, because I believe it will require some special treatment.

**Comment by X3N0-Life-Form** [comment]
Nice to have beam particles to go with our particle beams.

**Review by Goober5000 — APPROVED** [review]

**Pull request merged**